File: lambdas/stats-recorder/handler.go
// Lambda stats-recorder: actualiza estadísticas globales al terminar una partida.
//
// Trigger: EventBridge event "game.ended" publicado por round-ender.
//
// CONCEPTO GO — EventBridge vs SQS como trigger:
// EventBridge usa events.CloudWatchEvent (mismo tipo que CloudWatch Events).
// El campo Detail contiene el JSON que publicó round-ender.
// A diferencia de SQS, no hay batch — un evento = una invocación.
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"os"
	"sync"

	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/dynamodb"

	internaldynamo "github.com/quizarena/internal/dynamo"
)

var (
	logger   *slog.Logger
	dbClient *dynamodb.Client
	initOnce sync.Once
	initErr  error
)

func init() {
	logger = slog.New(slog.NewJSONHandler(os.Stdout, nil))
}

// gameEndedDetail es el payload del evento "game.ended" que publica round-ender.
type gameEndedDetail struct {
	RoomID  string         `json:"roomId"`
	Players []playerResult `json:"players"`
}

type playerResult struct {
	PlayerID string `json:"playerId"`
	Username string `json:"username"`
	Score    int    `json:"score"`
	IsWinner bool   `json:"isWinner"`
}

func handler(ctx context.Context, event events.CloudWatchEvent) error {
	if err := ensureClients(ctx); err != nil {
		return fmt.Errorf("init clients: %w", err)
	}

	// Deserializar el detail del evento EventBridge.
	var detail gameEndedDetail
	if err := json.Unmarshal(event.Detail, &detail); err != nil {
		return fmt.Errorf("unmarshal detail: %w", err)
	}

	logger.InfoContext(ctx, "recording game stats",
		slog.String("roomId", detail.RoomID),
		slog.Int("players", len(detail.Players)),
	)

	// Actualizar stats de cada jugador en paralelo.
	var wg sync.WaitGroup
	errs := make(chan error, len(detail.Players))

	for _, p := range detail.Players {
		wg.Add(1)
		go func(pr playerResult) {
			defer wg.Done()
			if err := internaldynamo.UpdatePlayerStats(ctx, dbClient, pr.PlayerID, pr.Username, pr.Score, pr.IsWinner); err != nil {
				errs <- fmt.Errorf("update stats for %s: %w", pr.PlayerID, err)
			}
		}(p)
	}
	wg.Wait()
	close(errs)

	// Recolectar errores — si alguno falló, retornamos error para que EventBridge reintente.
	for err := range errs {
		if err != nil {
			return err
		}
	}

	logger.InfoContext(ctx, "stats recorded", slog.String("roomId", detail.RoomID))
	return nil
}

func ensureClients(ctx context.Context) error {
	initOnce.Do(func() {
		cfg, err := config.LoadDefaultConfig(ctx)
		if err != nil {
			initErr = err
			return
		}
		dbClient = dynamodb.NewFromConfig(cfg)
	})
	return initErr
}

func main() {
	lambda.Start(handler)
}

File: lambdas/stats-recorder/handler.py
import json
import logging
from concurrent.futures import ThreadPoolExecutor, as_completed
import boto3

from shared.dynamo import stats as dynamo_stats

logger = logging.getLogger()
logger.setLevel(logging.INFO)

_db = None


def get_db():
    global _db
    if _db is None:
        _db = boto3.resource("dynamodb")
    return _db


def handler(event, context):
    """
    Trigger: EventBridge event "game.ended" publicado por round-ender.
    En Python, event['detail'] ya es un dict (no es raw JSON como en Go).
    """
    detail  = event["detail"]
    room_id = detail["roomId"]
    players = detail["players"]

    logger.info(json.dumps({"event": "recording-stats", "roomId": room_id, "players": len(players)}))

    db     = get_db()
    errors = []

    with ThreadPoolExecutor(max_workers=max(len(players), 1)) as executor:
        futures = {
            executor.submit(
                dynamo_stats.update_player_stats,
                db,
                p["playerId"],
                p["username"],
                p["score"],
                p["isWinner"],
            ): p
            for p in players
        }
        for future in as_completed(futures):
            try:
                future.result()
            except Exception as e:
                p = futures[future]
                logger.error(json.dumps({"event": "update-stats-failed", "playerId": p["playerId"], "error": str(e)}))
                errors.append(e)

    if errors:
        # Retornar error para que EventBridge reintente con backoff.
        raise RuntimeError(f"Fallaron {len(errors)} actualizaciones de stats")

    logger.info(json.dumps({"event": "stats-recorded", "roomId": room_id}))
